beangulp/importers/csv.py

```diff
@@ -149,6 +150,8 @@ def __init__(self, config, account, currency,
             narration fields of a source field.
           encoding: Encoding for the file, utf-8 if not specified or None.
           invert_sign: If true, invert the amount's sign unconditionally.
+          has_header: An optional boolean to indicate whether the csv file has a header.
+            If not set, the importer will try to detect it automatically.
```

Comment:
I don't particularly like the name of this argument. It is an argument passed to the importer constructor but the the think that has or not an header are the parsed files. I would much prefer a `names` that takes values `True` to force the first line of the file to be interpreted as column names, `False` to force it not to be interpreted as column names, `None` for auto detection as currently done, or a list of strings to do not interpret the first line of the file as column names but use the provided names instead.

Reply:
What would happen if the automatic detection detects a header but `use_names` is set to `False`? Would the first line still be read? If only column indexes are used, the variable would then only change whether the first line is skipped or not. So you could end up in the situation where you'd have to set `use_names=True` to skip the header even if you don't use names.

`skip_lines` could also be used for that purpose, but to me that's to skip a preceding section that is not arranged in columns.

Maybe a better solution is to never skip the header but require the user to account for it with `skip_lines` or if not set, set `skip_lines` in `normalize_config` to 1 or 0 depending on whether a header was detected.

Reply:
> What would happen if the automatic detection detects a header but `use_names` is set to `False`?

With `names` set to `False` there is no automatic detection.

> Maybe a better solution is to never skip the header but require the user to account for it with `skip_lines` or if not set, set `skip_lines` in `normalize_config` to 1 or 0 depending on whether a header was detected. That would fix the bug and wouldn't even require a new argument.

This would break backward compatibility and it is much clearer if`skip_lines` means skip not "maybe interpret them as column names".

Reply:
> > Maybe a better solution is to never skip the header but require the user to account for it with `skip_lines` or if not set, set `skip_lines` in `normalize_config` to 1 or 0 depending on whether a header was detected. That would fix the bug and wouldn't even require a new argument.
> 
> This would break backward compatibility and it is much clearer if`skip_lines` means skip not "maybe interpret them as column names".

I was probably not clear. That's what I meant, skip exactly `skip_lines` regardless of their interpretation. Currently it means 'skip `skip_lines` and then skip an additional line if a header is detected', which may fail if the detection is incorrect. That can be solved by overriding the detection or not skipping an extra line conditionally. It does break backward compatibility.
 
> 
> > What would happen if the automatic detection detects a header but `use_names` is set to `False`?
> 
> With `names` set to `False` there is no automatic detection.
> 

The bug is that the first line is incorrectly seen as a header and skipped in the file. I think a single-line CSV will just be interpreted as a header. The name `names` may be confusing because when there is a header but no names are used in the configuration, the setting should still be `names=True`, not `names=False` as some might expect. I think you intend `names` to mean 'the first line of this file contains the column names' and not 'I want to use names to refer to columns in my configuration', but to me the name isn't clear. If backwards compatibility can't be broken maybe there's another name for the argument.

Reply:
> The bug is that the first line is incorrectly seen as a header and skipped in the file. I think a single-line CSV will just be interpreted as a header. The name `names` may be confusing because when there is a header but no names are used in the configuration, the setting should still be `names=True`, not `names=False` as some might expect.

If you want the first line to be ignored just set `skip_lines=1`. I don't see what is confusing about it. `names=True` means use the first line after `skip_lines` for columns names. `names=False` means don't use the first line after `skip_lines` as column names. Obviously in this case you cannot have column definitions based on column names. `names=None` means autodetect if the first line after `skip_lines` has column names or data, like it is done unconditionally now. `names=[...]` mean use exactly the provided names for column names and don't do anything special with the first line after `skip_lines`. This last possible value for the option would be nice to have and not hard to implement, but I would not insist on it.

The name of the option and its interpretation is based on the `numpy.genfromtxt` function.

Another possibility would be to have an `header` parameter. `header=None` means autodetect if the first line after `skip_lines` has column names or data, like it is done unconditionally now. `header=N` means use line N for columns names, ignore all lines up to that one. `header=False` means there is not line with column names. Obviously in this case you cannot have column definitions based on column names. This has the drawback that the interaction with `skip_lines` is ill defined: should `skip_lines=5, header=3` use line 8 for the column names or use line 3 for the header and start reading data at line 8?

This would be loosely modeled after the `pandas.read_csv` function. But Pandas is not the best source of inspiration for clear and concise APIs IMHO. I think the other alternative fits better.